### DISCUSSION TITLE
Running nodejs (cjs) libs with node_modules deps using `deno-runtime`

### DISCUSSION THREAD

**Discussion by rouzwelt** [denoland/deno#23468] [Q&A] [answered]
Hi, so I 've been trying to use deno-runtime to run some js code from rust, the js code has some node_modules deps as well as nodejs builtins and is in cjs format, so I have a few questions about this that someone may help me with:

do I need to pull deno-std from remote as a side-module when I instantiated the `MainWorker` or is there some sort of an option on `MainWorker` that will basically do this i.e. preload the nodejs builtins? because I was able to pull in the deno-std and run the mentioned lib in esm format with nodejs builtins working fine, but I was not able to work it out with cjs and when `require` is involved. so I get some error ` Deno.core.evalContext is not supported in this environment` when I import `createRequire` and then require other js module as you use a require statement normally in cjs.

so I think Im pretty close, I was able to work it out with esm format, by pulling in the `deno-std/node/module.ts`, however Im not sure if that's a necessary step, I mean if there is an option on `MainWorker` that does this without me preloading the deno-sdt/node manually. 

Im pretty new to deno, so I would appreciate if anyone can help me. thanks

**Comment by bartlomieju** [comment]
You shouldn't use `deno_std/node` - it's been long deprecated and merged into the main runtime. By default `require` is not exposed to user code - you can create one manually like so:
```
import { createRequire } from "node:module";
const require = createRequire(import.meta.url);
```

And then you can assign it to `globalThis` to make it available globally: `globalThis.require = require`. Note that it's somewhat of a hack and will probably not work in all situations.

**Reply by rouzwelt** [reply]
awesome thanks for clarification, so is this correct assumption that using cjs libs imported in main the module (loading them from node_modules) might not always work correctly as they might have require statements in them that might not work?
and one other question is that, besides this hack for `require` that you say might not always work, is there any other issues with laoding cjs modules that one needs to know about? I mean I checked the `deno cli` and found an algo that converts cjs to esm, is that some sort of unstable feature too when importing cjs modules with esm import statements? 
well I can ask it in some simpler way, that is loading and using cjs modules reliable atm without considering the `require` hack you said might not always work?

**Reply by bartlomieju** [reply, marked as answer]
> so is this correct assumption that using cjs libs imported in main the module (loading them from node_modules) might not always work correctly as they might have require statements in them that might not work?

That's correct.

> and one other question is that, besides this hack for require that you say might not always work, is there any other issues with laoding cjs modules that one needs to know about? I mean I checked the deno cli and found an algo that converts cjs to esm, is that some sort of unstable feature too when importing cjs modules with esm import statements?

That code is only required when an ESM module tries to `import` a CJS module. If you want to execute only CJS then it shouldn't be needed.
